fast_llm/engine/training/config.py

```diff
@@ -90,71 +141,109 @@ class WandbConfig(Config):
+            iterations = [iteration for iteration in iterations if self.get_count(iteration) % self.keep_every != 0]
+        # Exclude the last `keep`.
+        return iterations[: -self.keep]
 
 
 @config_class()
-class CallbackConfig(Config):
-    script: list[str] | None = Field(
-        default=None,
-        desc="Shell script to run after.",
-        hint=FieldHint.feature,
-        valid=skip_valid_if_none(_validate_script),
-    )
-    environment: dict[str, str] = Field(
-        default_factory=dict,
-        desc="Environment variables to add to the script.",
-        hint=FieldHint.feature,
+class CheckpointConfig(CheckpointBaseConfig):
+    _abstract = False
+    save_name: typing.ClassVar[str] = "checkpoint"
+    # TODO v0.2: Rename to `checkpoint` so we don't need this extra variable?
+    directory_name = "checkpoints"
+    interval = FieldUpdate(
+        desc="The number of training iterations between each checkpoint." " Setting to None will disable checkpoints."
     )
-
-    def run(self):
-        if self.script is not None:
-            environment = os.environ.copy()
-            environment.update(self.environment)
-            subprocess.Popen(self.script, env=environment)
+    offset = FieldUpdate(desc="Offset for the first checkpoint.")
+    callback: CallbackConfig = FieldUpdate(desc="Callback (shell script) to run after checkpoint.")
+    keep: int | None = FieldUpdate(default=5)
+
+    def get_save_config(self, path: pathlib.Path):
+        return CheckpointSaveConfig(
+            path=path,
+            format=CheckpointFormat.distributed,
+            model_weights=True,
+            optimizer_state=True,
+        )
 
 
 @config_class()
-class ExportConfig(get_interval_config_class("export")):
-    callback: CallbackConfig = Field(
-        default_factory=CallbackConfig,
-        desc="Callback (shell script) to run after export.",
-        hint=FieldHint.core,
+class ExportConfig(CheckpointBaseConfig, CheckpointConfigBase, CheckpointStateConfigBase, CheckpointSaveConfigBase):
+    _abstract = False
+    save_name: typing.ClassVar[str] = "export"
+    directory_name = "export"
+    interval = FieldUpdate(
+        desc="The number of training iterations between each export." " Setting to None will disable exports."
     )
+    offset = FieldUpdate(desc="Offset for the first export.")
+    callback: CallbackConfig = FieldUpdate(desc="Callback (shell script) to run after export.")
+
+    def get_save_config(self, path: pathlib.Path):
+        return CheckpointSaveConfig.from_dict(self, {"path": path}, strict=False)
 
 
 @config_class()
-class ShutdownConfig(get_interval_config_class("automated shutdown")):
-    pass
+class ShutdownConfig(IntervalConfig):
+    interval = FieldUpdate(
+        desc="The number of training iterations between each automated shutdown."
+        " Setting to None will disable automated shutdowns."
+        " Must be a sub-interval of the checkpoint interval."
```

Comment:
what's an automated shutdown? I am not familiar with this feature? is this new?

Reply:
It's always been there but never used.